### PR TITLE
[fix]: [CDS-85196]: Bulk Triggers API

### PR DESCRIPTION
**Pull request by vtxorxwitty**
## Describe your changes

New API for Enabling/Disabling triggers in bulk. 

Design Spec - https://harness.atlassian.net/browse/CDS-85196?focusedCommentId=619097
Test Plan - https://harness.atlassian.net/browse/CDS-85196?focusedCommentId=621078

API response -  
Screenshot 2023-12-06 at 4 07 57 PM

Triggers getting disabled on UI - 
Screenshot 2023-12-07 at 9 16 17 AM

Yaml getting updated as well- 
Screenshot 2023-12-07 at 9 16 31 AM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56250)

### CODE REVIEW

**Comment by vtxorxwitty** [comment]
trigger uts

**Comment by vtxorxwitty** [comment]
trigger uts

**Comment by vtxorxwitty** [comment]
trigger uts

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56250)* status: 13 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/ng-triggers/contracts/src/main/java/io/harness/ngtriggers/resource/NGTriggerResource.java` line 378 at r4](https://reviewable.io/reviews/harness/harness-core/56250#-NkwAcKS9mICR6DaJDd1:-NkwAcKS9mICR6DaJDd2:ba7bjft) ([raw file](https://github.com/harness/harness-core/blob/7f2974794e3667d6ebdd6584e32b3d3c10b2d2f2/pipeline-service/modules/ng-triggers/contracts/src/main/java/io/harness/ngtriggers/resource/NGTriggerResource.java#L378)):*
> ```Java
> 
>   @PATCH
>   @Path("/bulk")
> ```

The path doesn't look indicative of what it does. Shouldn't it be more like update-triggers-bulk or bulk/update?

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56250)* status: 13 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/mapper/TriggerFilterHelper.java` line 286 at r4](https://reviewable.io/reviews/harness/harness-core/56250#-NkwBMDyBbwNojqjd8iD:-NkwBMDyBbwNojqjd8iE:bk74l09) ([raw file](https://github.com/harness/harness-core/blob/7f2974794e3667d6ebdd6584e32b3d3c10b2d2f2/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/mapper/TriggerFilterHelper.java#L286)):*
> ```Java
>     }
>     if (StringUtils.isBlank(orgIdentifier) && StringUtils.isNotBlank(projectIdentifier)) {
>       throw new InvalidRequestException(
> ```

do we need such combinational validation checks? Can't we simply put null checks for each field?

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56250)* status: 13 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/mapper/TriggerFilterHelper.java` line 286 at r4](https://reviewable.io/reviews/harness/harness-core/56250#-NkwBMDyBbwNojqjd8iD:-NkwBULJ9IP8i4HunO6l:bpavmtw) ([raw file](https://github.com/harness/harness-core/blob/7f2974794e3667d6ebdd6584e32b3d3c10b2d2f2/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/mapper/TriggerFilterHelper.java#L286)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

do we need such combinational validation checks? Can't we simply put null checks for each field?
</blockquote></details>

I'm assuming here that triggers exist only on project level ^^

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56250)* status: 13 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/ng-triggers/contracts/src/main/java/io/harness/ngtriggers/resource/NGTriggerResource.java` line 378 at r4](https://reviewable.io/reviews/harness/harness-core/56250#-NkwAcKS9mICR6DaJDd1:-NkwD5AD5mCsDaa6wUur:buwa8sk) ([raw file](https://github.com/harness/harness-core/blob/7f2974794e3667d6ebdd6584e32b3d3c10b2d2f2/pipeline-service/modules/ng-triggers/contracts/src/main/java/io/harness/ngtriggers/resource/NGTriggerResource.java#L378)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

The path doesn't look indicative of what it does. Shouldn't it be more like update-triggers-bulk or bulk/update?
</blockquote></details>

Based on the PR, it seems it only enables / disables triggers in bulk. So then better path would be "bulk/toggle" or "toggle-triggers-bulk"

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56250)* status: 13 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java` line 1483 at r4](https://reviewable.io/reviews/harness/harness-core/56250#-NkwDJ9D7lchiL-iQEQ3:-NkwDJ9D7lchiL-iQEQ4:bgki6pa) ([raw file](https://github.com/harness/harness-core/blob/7f2974794e3667d6ebdd6584e32b3d3c10b2d2f2/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java#L1483)):*
> ```Java
> 
>   @Override
>   public TriggerUpdateCount enablingDisablingTriggersInBulk(
> ```

better name could be "toggleTriggersInBulk"

**Review by vtxorxwitty — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56250)* status: 13 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/ng-triggers/contracts/src/main/java/io/harness/ngtriggers/resource/NGTriggerResource.java` line 378 at r4](https://reviewable.io/reviews/harness/harness-core/56250#-NkwAcKS9mICR6DaJDd1:-NkwlGg20p3TmBqKHbtc:b-70vigb) ([raw file](https://github.com/harness/harness-core/blob/7f2974794e3667d6ebdd6584e32b3d3c10b2d2f2/pipeline-service/modules/ng-triggers/contracts/src/main/java/io/harness/ngtriggers/resource/NGTriggerResource.java#L378)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

Based on the PR, it seems it only enables / disables triggers in bulk. So then better path would be "bulk/toggle" or "toggle-triggers-bulk"
</blockquote></details>

bulk/toggle looks more indicative. Will update that.
___
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/mapper/TriggerFilterHelper.java` line 286 at r4](https://reviewable.io/reviews/harness/harness-core/56250#-NkwBMDyBbwNojqjd8iD:-NkwlQEyAqoLiwUYDnrJ:b-2lhcek) ([raw file](https://github.com/harness/harness-core/blob/7f2974794e3667d6ebdd6584e32b3d3c10b2d2f2/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/mapper/TriggerFilterHelper.java#L286)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

I'm assuming here that triggers exist only on project level ^^
</blockquote></details>

@mohitg0795 the aim for this API is to toggle all triggers for diff scopes like account/org/..
We cannot add null checks as you'll have to pass them for every API call. But we don't need the every time. For example, 
If we want to disable all triggers for an organization "abc", the projectIdentifier sent in the API will be null and hence the DB query criteria will be formed. 
___
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java` line 1483 at r4](https://reviewable.io/reviews/harness/harness-core/56250#-NkwDJ9D7lchiL-iQEQ3:-NkwmC0W8sy0c73FlUVP:bdf7zgi) ([raw file](https://github.com/harness/harness-core/blob/7f2974794e3667d6ebdd6584e32b3d3c10b2d2f2/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java#L1483)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

better name could be "toggleTriggersInBulk"
</blockquote></details>

Will update this.

**Comment by vtxorxwitty** [comment]
trigger uts

**Comment by vtxorxwitty** [comment]
trigger uts

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/repositories/custom/NGTriggerRepositoryCustomImpl.java [hunk 1]
```diff
@@ -256,6 +256,20 @@ public boolean updateManyTriggerPollingSubscriptionStatusBySignatures(String acc
     return true;
   }
 
+  @Override
+  public long toggleTriggersInBulk(String accountId, boolean enable, Criteria criteria) {
```
Comment (rathodmeetsatish) [inline pipeline-service/modules/ng-triggers/src/main/java/io/harness/repositories/custom/NGTriggerRepositoryCustomImpl.java:260]: Here we are just updating the enabled field but we should also update yaml, we already have a method to disable trigger in bulk io.harness.ngtriggers.service.impl.NGTriggerServiceImpl#disableTriggers
We can modify this(disableTriggers) method itself to enable/disable triggers in bulk

**Review by rathodmeetsatish — CHANGES_REQUESTED** [review]

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/repositories/custom/NGTriggerRepositoryCustomImpl.java (hunk 1 above)
Comment (vtxorxwitty) [reply pipeline-service/modules/ng-triggers/src/main/java/io/harness/repositories/custom/NGTriggerRepositoryCustomImpl.java:260]: Checked that method already @rathodmeetsatish. The ask here is a bit different and we want it for only Admin users behind the access control. Also Yaml updation is not the ask as we just want an API to toggle the trigger.

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java [hunk 2]
```diff
@@ -581,43 +585,55 @@ public NGTriggerEntity update(NGTriggerEntity ngTriggerEntity, NGTriggerEntity o
...
-    criteria.and(NGTriggerEntityKeys.deleted).is(false);
-    CloseableIterator<NGTriggerEntity> iterator = ngTriggerRepository.findAll(criteria);
-    List<NGTriggerEntity> toBeDisabledTriggers = new ArrayList<>();
+  public TriggerUpdateCount toggleTriggers(
+      boolean enable, String accountIdentifier, String orgIdentifier, String projectIdentifier) {
+    Criteria criteria = TriggerFilterHelper.getCriteriaForTogglingTriggersInBulk(
+        enable, accountIdentifier, orgIdentifier, projectIdentifier, null, null);
+
+    List<NGTriggerEntity> triggerEntities =
+        list(criteria, PageRequest.of(0, 100000, Sort.by(Sort.Direction.DESC, NGTriggerEntityKeys.createdAt)))
```
Comment (rathodmeetsatish) [inline pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java:594]: why are we using this and not io.harness.repositories.custom.NGTriggerRepositoryCustomImpl#findAll(org.springframework.data.mongodb.core.query.Criteria) method?

**Review by rathodmeetsatish — CHANGES_REQUESTED** [review]

**Comment by vtxorxwitty** [comment]
trigger ut5, ut6, ut7, ut8, ut9

**Review by mohitg0795 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56250)* status: 17 file(s) to review. 2 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/mapper/TriggerFilterHelper.java` line 286 at r4](https://reviewable.io/reviews/harness/harness-core/56250#-NkwBMDyBbwNojqjd8iD:-Nl2Y8o1EVs-ZGe6nkWX:b-4hvi0m) ([raw file](https://github.com/harness/harness-core/blob/7f2974794e3667d6ebdd6584e32b3d3c10b2d2f2/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/mapper/TriggerFilterHelper.java#L286)):*
<details><summary><i>Previously, vtxorxwitty (Ved Tambat) wrote…</i></summary><blockquote>

@mohitg0795 the aim for this API is to toggle all triggers for diff scopes like account/org/..
We cannot add null checks as you'll have to pass them for every API call. But we don't need the every time. For example, 
If we want to disable all triggers for an organization "abc", the projectIdentifier sent in the API will be null and hence the DB query criteria will be formed.
</blockquote></details>

Got it.

**Review by vcalasansh — APPROVED** [review]
> Reviewed 1 of 11 files at r1, 1 of 3 files at r2, 2 of 11 files at r3, 1 of 7 files at r4, 3 of 9 files at r7, 4 of 9 files at r8, 1 of 2 files at r9, 4 of 4 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56250)* status: 17 file(s) to review. 2 change requests. 2. unresolved discussions. 1 of 2 approvals obtained.

**Pull request merged**